### PR TITLE
Project, Commission, and Unclog Nearline Scanners

### PR DESCRIPTION
**Pull request by Dala0**
Adds a new scanner which records data from Vidispine collections that are flagged as either projects or commissions.

Also adds a new scanner that loads Vidispine file data, attempts to load Vidispine item data, then uses the metadata from the item data to load data about which projects the item is in. This data is then used to create an index of data about what state the file should be in based on the status of the projects it is a member of and other flags set in the metadata.

Tested on a machine running macOS 10.12.6 and a Kubernetes cluster.

@fredex42

### CODE REVIEW

File: common/src/main/scala/models/PlutoProject.scala [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+import scala.util.Try
+
+case class PlutoProject (siteIdentifier:Option[String], collectionId:Int, name:Option[String], status:String, commissionCollection:Option[Int]) {
+  def commissionVSID() = s"${siteIdentifier.getOrElse("VX")}-$commissionCollection"
+  def projectVSID() = s"${siteIdentifier.getOrElse("VX")}-$collectionId"
+}
+
+
+/*
```
Comment (fredex42) [inline common/src/main/scala/models/PlutoProject.scala]: can remove this commented json

File: common/src/main/scala/models/UnclogOutput.scala [hunk 2]
```diff
@@ -0,0 +1,15 @@
+package models
+import io.circe.{Encoder, Decoder}
+
+object MediaStatusValue extends Enumeration {
+  val SHOULD_BE_ARCHIVED,NO_PROJECT,BRANDING,DELETABLE,SENSITIVE,PROJECT_HELD,PROJECT_OPEN,UNKNOWN,NO_ITEM = Value
+  //PROJECT_OPEN_COMMISSION_COMPLETED,PROJECT_OPEN_COMMISSION_EXPIRED,PROJECT_OPEN_COMMISSION_OPEN
```
Comment (fredex42) [inline common/src/main/scala/models/UnclogOutput.scala]: can remove commented code

File: common/src/main/scala/models/VidispineProject.scala [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+case class VidispineProject (vsid:String, name:Option[String], status:Option[String], commissionCollection:Option[String], deepArchive:Boolean, deletable:Boolean, sensitive:Boolean, created:Option[ZonedDateTime], modified:Option[ZonedDateTime])
+
+object VidispineProject {
+  def extractMetadataValue(metaNode:Node, fieldName:String):Option[String] = {
+    val applicableFieldNodes = (metaNode \ "timespan" \ "field").filter(fieldNode=>(fieldNode\"name").text==fieldName)
+    applicableFieldNodes.headOption.map(fieldNode=>(fieldNode\"value").text)
+  }
+
+  def checkField(field:Option[String], check:Option[String]): Boolean = {
```
Comment (fredex42) [inline common/src/main/scala/models/VidispineProject.scala]: can simplify this method to `def checkField(field:Option[String], check:Option[String]): Boolean = field==check`

File: common/src/main/scala/vidispine/VSLazyItem.scala [hunk 4]
```diff
@@ -69,6 +79,7 @@ case class VSLazyItem (itemId:String, lookedUpMetadata:Map[String,VSMetadataEntr
 }
 
 object VSLazyItem extends ((String,Map[String,VSMetadataEntry],Option[Map[String,VSShape]])=>VSLazyItem) {
+  private val logger = LoggerFactory.getLogger(getClass)
```
Comment (fredex42) [inline common/src/main/scala/vidispine/VSLazyItem.scala]: can take this out again if nothing is using it

File: common/src/test/scala/models/PlutoProjectSpec.scala [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+                      |      80
+                      |    ],
+                      |    "gnm_project_project_file_item": null,
+                      |    "gnm_project_prelude_file_item": null,
+                      |    "gnm_project_type": "330b4d84-ef24-41e2-b093-0d15829afa64",
+                      |    "project_locker_id": 6053,
+                      |    "project_locker_id_prelude": null
+                      |  }""".stripMargin
+      val a = io.circe.parser.parse(rawJson).flatMap(_.as[PlutoProject])
+      a
```
Comment (fredex42) [inline common/src/test/scala/models/PlutoProjectSpec.scala]: can take this line out

File: unclognearline/src/main/scala/streamcomponents/DeletableSwitch.scala [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+import akka.stream.stage.{AbstractInHandler, AbstractOutHandler, GraphStage, GraphStageLogic}
+import akka.stream.{Attributes, Inlet, Outlet, UniformFanOutShape}
+import models.UnclogStream
+import org.slf4j.LoggerFactory
+
+/**
+  * Checks whether the incoming object contains a VidispineProject object set to deletable.
+  * If so, pushes it to the outYes port, if not pushes it to the outNo port.
+  */
+class DeletableSwitch extends GraphStage[UniformFanOutShape[UnclogStream, UnclogStream]]{
```
Comment (fredex42) [inline unclognearline/src/main/scala/streamcomponents/DeletableSwitch.scala]: we have a problem here; thinking about it carefully, we want to match "Deep Archive" if _any_ project is "deepArchive" (as above), but we only want to match "Deletable" if _all_ projects are "deletable".  Please invert the check at line 35 (i.e., push to No immediately and return if the project is _not_ deletable)

File: unclognearline/src/main/scala/streamcomponents/ItemSwitch.scala [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+      } catch {
+        case err:Throwable=>
+          logger.error("Uncaught exception checking for an item: ", err)
+          failStage(err)
+      }
+
+      def onPushBody(): Unit = {
+        val elem = grab(in)
+
+        if (elem.VSItem.isDefined) {
```
Comment (fredex42) [inline unclognearline/src/main/scala/streamcomponents/ItemSwitch.scala]: not massively elegant but it doesn't matter :)

File: unclognearline/src/test/scala/ProjectCountSwitchSpec.scala [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+      val pattern = DateTimeFormatter.ISO_OFFSET_DATE_TIME
+
+      val testStream = GraphDSL.create(sinkFactory) { implicit builder=> sink=>
+        import akka.stream.scaladsl.GraphDSL.Implicits._
+
+        val src = builder.add(Source.single(UnclogStream(VSFile("VX-123344", "/test", "/test", None, 2318793, None, ZonedDateTime.parse("2019-10-16T04:00:00.523+01:00", pattern), 1, "VX-4", None, None, None, None),None,Seq(VidispineProject("VX-1234", Option("Test Name"), Option("New"), Option("VX-12"), true, false, true, Option(ZonedDateTime.parse("2019-07-17T10:35:12.598Z", pattern)), Option(ZonedDateTime.parse("2019-10-16T04:00:00.523+01:00", pattern)))), None)))
+        val switch = builder.add(new ProjectCountSwitch)
+        val merge = builder.add(Merge[String](2))
+        src ~> switch
+        switch.out(0).map(rpt=>"YES") ~> merge //YES branch
```
Comment (fredex42) [inline unclognearline/src/test/scala/ProjectCountSwitchSpec.scala]: like this :+1:

File: collectionscanner/src/main/scala/CommissionScanner.scala [hunk 9]
```diff
@@ -0,0 +1,254 @@
...
+        case _ => System.exit(exitCode)
+      })
+    })
+  }
+
+  /**
+    * Builds the main stream for conducting the project scan
+    * @return
+    */
+  def buildProjectStream(initialJobRecord:JobHistory)(implicit jobHistoryDAO: JobHistoryDAO, esClient:ElasticClient,
```
[the remark this answers is not in the document]
Comment (Dala0) [reply collectionscanner/src/main/scala/CommissionScanner.scala]: Done.

File: collectionscanner/src/main/scala/CommissionScanner.scala [hunk 10]
```diff
@@ -0,0 +1,254 @@
...
+      case Right(results)=>
+        logger.info(s"Cleanout of old census jobs succeeded: $results")
+    }), 5 minutes)
+
+    val runInfo = JobHistory.newRun(JobType.CommissionScan)
+
+    val resultFuture = jobHistoryDAO.put(runInfo).flatMap({
+      case Right(_)=>
+        logger.info(s"Saved run info ${runInfo.toString}")
+        val commissionStreamFut = RunnableGraph.fromGraph(buildStream(runInfo)).run()//.map(resultCount=>Right(resultCount))
```
[the remark this answers is not in the document]
Comment (Dala0) [reply collectionscanner/src/main/scala/CommissionScanner.scala]: Done.

File: collectionscanner/src/main/scala/streamcomponents/PlutoProjectSource.scala [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+
+import akka.stream.{Attributes, Materializer, Outlet, SourceShape}
+import akka.stream.stage.{AbstractOutHandler, GraphStage, GraphStageLogic}
+import models.PlutoProject
+import org.slf4j.LoggerFactory
+import vidispine.VSCommunicator
+import scala.util.{Failure, Success}
+import scala.concurrent.ExecutionContext.Implicits._
+
+class PlutoProjectSource(recordsPerPage:Int=100)(implicit comm:VSCommunicator,mat:Materializer) extends GraphStage[SourceShape[PlutoProject]] {
```
[the remark this answers is not in the document]
Comment (Dala0) [reply collectionscanner/src/main/scala/streamcomponents/PlutoProjectSource.scala]: Done.

File: collectionscanner/src/main/scala/streamcomponents/VidispineProjectSource.scala [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+import akka.stream.stage.{AbstractOutHandler, GraphStage, GraphStageLogic}
+import akka.stream.{Attributes, Materializer, Outlet, SourceShape}
+import models.VidispineProject
+import org.slf4j.LoggerFactory
+import vidispine.VSCommunicator
+
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.util.{Failure, Success}
+
+class VidispineProjectSource(recordsPerPage:Int=5)(implicit comm:VSCommunicator, mat:Materializer) extends GraphStage[SourceShape[VidispineProject]] {
```
[the remark this answers is not in the document]
Comment (Dala0) [reply collectionscanner/src/main/scala/streamcomponents/VidispineProjectSource.scala]: Done.

File: common/src/main/scala/models/PlutoCommission.scala [hunk 13]
```diff
@@ -0,0 +1,175 @@
...
+    applicableFieldNodes.headOption.map(fieldNode=>(fieldNode\"value").text)
+  }
+
+  def extractProjectList(metaNode:Node):List[String] = {
+    val projectIdNodes = (metaNode \ "timespan" \ "field").filter(fieldNode=>(fieldNode\"name").text=="__child_collection")
+    projectIdNodes.map(node=>(node\"value").text).toList
+  }
+
+  /*
+<collection>
```
[the remark this answers is not in the document]
Comment (Dala0) [reply common/src/main/scala/models/PlutoCommission.scala]: Done.

File: common/src/main/scala/models/PlutoProject.scala (hunk 1 above)
Comment (Dala0) [reply common/src/main/scala/models/PlutoProject.scala]: Done.

File: common/src/main/scala/models/UnclogOutput.scala (hunk 2 above)
Comment (Dala0) [reply common/src/main/scala/models/UnclogOutput.scala]: Done.

File: common/src/main/scala/models/VidispineProject.scala (hunk 3 above)
Comment (Dala0) [reply common/src/main/scala/models/VidispineProject.scala]: Done.

File: common/src/main/scala/vidispine/VSLazyItem.scala (hunk 4 above)
Comment (Dala0) [reply common/src/main/scala/vidispine/VSLazyItem.scala]: Done.

File: common/src/test/scala/models/PlutoProjectSpec.scala (hunk 5 above)
Comment (Dala0) [reply common/src/test/scala/models/PlutoProjectSpec.scala]: Done.

File: unclognearline/src/main/scala/streamcomponents/DeletableSwitch.scala (hunk 6 above)
Comment (Dala0) [reply unclognearline/src/main/scala/streamcomponents/DeletableSwitch.scala]: I wrote this file before you told me we did not need it so I am just going to delete it as it is not used.

Comment (fredex42) [reply unclognearline/src/main/scala/streamcomponents/DeletableSwitch.scala]: ok

File: common/src/main/scala/vidispine/VSLazyItem.scala (hunk 4 above)
Comment (Dala0) [reply common/src/main/scala/vidispine/VSLazyItem.scala]: Removing this broke something so I put it back.

**Pull request merged**